### PR TITLE
Removes write consistency level across replication action APIs in favor of wait_for_active_shards

### PR DESCRIPTION
**Pull request by abeyad**
Removes the notion of write consistency level across replication action APIs 
(index, update, delete, re-index, etc) in favor of waiting for active shard copy 
count (wait_for_active_shards) that was introduced in #18985 for index creation.

The reason for the removal of write consistency for replication operations is that 
it led to a false notion of what guarantees were made by checking the "write 
consistency" level.  The way write consistency checks worked before was to check 
the number of active shard copies as a replication request (e.g. indexing) arrived.  If 
the required number of active shards were present (defaulted to "quorum"),
then the replication operation would proceed.  However, it is entirely possible that
replication would still fail on a majority or even all replicas after the write consistency
check, but the operation still succeeds on the primary.  Also, our use of the terms 
"write consistency" and "quorum" in relation to replication actions did not convey 
the same semantics as found in the literature, as we don't have an equivalent 
notion of read consistency and quorum typically relates to a guarantee in the 
consensus model which we do not have for replication actions.  Hence, we have 
changed the terminology to use `wait_for_active_shards` as it better conveys the fact 
that this check is a best attempt at maintaining the desired level of replication and 
thereby resiliency.

This PR also removes the `index.write_consistency` setting in favor of a new setting: `index.write.wait_for_active_shards`.  `index.write.wait_for_active_shards` is an 
index-level setting that is dynamically updatable, and can be set to either `all` or 
any non-negative value up to the total number of shard copies for a shard in an 
index (number of replicas + 1).  The default value for this setting is 1 (meaning only 
wait for the primary shard).

### CODE REVIEW

**Comment by abeyad** [comment]
@bleskes @ywelsch FYI, I marked it as a WIP in case we want to change the name from wait_for_active_shards to something else.  Otherwise its ready for review.

File: core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java [hunk 1]
```diff
@@ -167,12 +167,10 @@ public boolean enoughShardsActive(final ClusterState clusterState, final String
      * Returns true iff the active shard count in the shard routing table is enough
      * to meet the required shard count represented by this instance.
      */
-    public boolean enoughShardsActive(final IndexShardRoutingTable shardRoutingTable, final IndexMetaData indexMetaData) {
-        if (shardRoutingTable.activeShards().size() < resolve(indexMetaData)) {
-            // not enough active shard copies yet
-            return false;
-        }
-        return true;
+    public EvalResult enoughShardsActive(final IndexShardRoutingTable shardRoutingTable, final IndexMetaData indexMetaData) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java]: can we try to avoid passing in the indexMetaData? I think the shard routing table should have all we need. The resolve method can be changed, or even removed as we can special case the testing for all (check for unassigned) and none (allways succeed) here.

Also, it would be great if we can remove the EvalResult class. We currently create it for each operation. Instead I would suggest having the boolean as we had before, and people can either call a describe method to get a string description or construct a string themselves. All the info they need is easily available, with the exception of the resolve() value, but that's clear from the name of the ActiveShardCount.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 2]
```diff
@@ -80,10 +81,10 @@
 
     public ReplicationOperation(Request request, Primary<Request, ReplicaRequest, PrimaryResultT> primary,
                                 ActionListener<PrimaryResultT> listener,
-                                boolean executeOnReplicas, boolean checkWriteConsistency,
+                                boolean executeOnReplicas, boolean checkActiveShardCount,
                                 Replicas<ReplicaRequest> replicas,
                                 Supplier<ClusterState> clusterStateSupplier, ESLogger logger, String opType) {
-        this.checkWriteConsistency = checkWriteConsistency;
+        this.checkActiveShardCount = checkActiveShardCount;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: can we see check and see if we can get rid of this boolean infavour of setting the request waitForActiveShards to NONE by default on all the relevant requests?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 3]
```diff
@@ -122,7 +123,7 @@ protected TransportReplicationAction(Settings settings, String actionName, Trans
 
         this.transportOptions = transportOptions();
 
-        this.defaultWriteConsistencyLevel = WriteConsistencyLevel.fromString(settings.get("action.write_consistency", "quorum"));
+        this.defaultWaitForActiveShards = IndexSettings.WAIT_FOR_ACTIVE_SHARDS_SETTING.get(settings);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: can we make this dynamic while at it?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 4]
```diff
@@ -165,9 +166,9 @@ protected void resolveRequest(MetaData metaData, IndexMetaData indexMetaData, Re
     protected abstract ReplicaResult shardOperationOnReplica(ReplicaRequest shardRequest);
 
     /**
-     * True if write consistency should be checked for an implementation
+     * True if the active shard count should be checked before proceeding with the replication action.
      */
-    protected boolean checkWriteConsistency() {
+    protected boolean checkActiveShardCount() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: see comment above about trying to get rid of this method.

File: core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java [hunk 5]
```diff
@@ -68,7 +69,7 @@
     List<Object> payloads = null;
 
     protected TimeValue timeout = BulkShardRequest.DEFAULT_TIMEOUT;
-    private WriteConsistencyLevel consistencyLevel = WriteConsistencyLevel.DEFAULT;
+    private ActiveShardCount activeShardCount = ActiveShardCount.DEFAULT;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: can we call those waitForActiveShards, like the methods?

File: core/src/test/java/org/elasticsearch/search/basic/SearchWhileCreatingIndexIT.java [hunk 6]
```diff
@@ -54,11 +54,8 @@ public void testTwoReplicas() throws Exception {
 
     private void searchWhileCreatingIndex(boolean createIndex, int numberOfReplicas) throws Exception {
 
-        // make sure we have enough nodes to guaranty default QUORUM consistency.
-        // TODO: add a smarter choice based on actual consistency (when that is randomized)
-        int shardsNo = numberOfReplicas + 1;
-        int neededNodes = shardsNo <= 2 ? 1 : shardsNo / 2 + 1;
-        internalCluster().ensureAtLeastNumDataNodes(randomIntBetween(neededNodes, shardsNo));
+        // TODO: add a smarter choice based on varying active shard count (when that is randomized)
+        internalCluster().ensureAtLeastNumDataNodes(randomIntBetween(1, numberOfReplicas + 1));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/search/basic/SearchWhileCreatingIndexIT.java]: do we need this with the new default waitForActiveShards?

File: core/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java [hunk 7]
```diff
@@ -690,7 +690,7 @@ public void sendResponse(RestResponse response) {
     @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/12621")
     public void testChaosSnapshot() throws Exception {
         final List<String> indices = new CopyOnWriteArrayList<>();
-        Settings settings = Settings.builder().put("action.write_consistency", "one").build();
+        Settings settings = Settings.builder().put(IndexSettings.WAIT_FOR_ACTIVE_SHARDS_SETTING.getKey(), "1").build();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java]: this is the default no?

File: rest-api-spec/src/main/resources/rest-api-spec/api/delete.json [hunk 8]
```diff
@@ -23,10 +23,9 @@
         }
       },
       "params": {
-        "consistency": {
-          "type" : "enum",
-          "options" : ["one", "quorum", "all"],
-          "description" : "Specific write consistency setting for the operation"
+        "wait_for_active_shards": {
+          "type" : "string",
+          "description" : "Explicit active shard count required for the operation"
```
Comment (bleskes) [inline rest-api-spec/src/main/resources/rest-api-spec/api/delete.json]: can we copy over the descriptions from the java docs? we should make it clear this is required for the operation to start.

**Comment by bleskes** [comment]
Thx @abeyad . Left some comments. I also miss the (migration) doc changes. Are those coming in another PR (and please, let's not forget :))

File: core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java (hunk 1 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java]: @bleskes the primary motivation for the `resolve` method was to abstract the value within the class.  you are correct in that we could just pass in the number of replicas instead of the entire `IndexMetaData`, but I believe the `resolve` method has utility, otherwise we have to expose a `getValue` method on `ActiveShardCount`, which seems easily avoided if we keep the `resolve` method.  Thoughts?

File: core/src/test/java/org/elasticsearch/search/basic/SearchWhileCreatingIndexIT.java (hunk 6 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/search/basic/SearchWhileCreatingIndexIT.java]: correct we don't need this

**Comment by abeyad** [comment]
@bleskes 
https://github.com/elastic/elasticsearch/pull/19454/commits/c23e9da971f07066b6e7c223148784c73041d7f2 addresses the code review in general 
https://github.com/elastic/elasticsearch/pull/19454/commits/536a2dbb81e00f81605b81cf155dd0016ee50ccc removes the checkWriteConsistency() method from transport replication actions in favor of `ActiveShardCount`
https://github.com/elastic/elasticsearch/pull/19454/commits/762301f69a00f5110efb2ce606a647a392e0fed1 makes the `index.write.wait_for_active_shards` setting updatable

I have a couple things to discuss with you regarding documentation, then I will push that up in this PR as well.

**Comment by abeyad** [comment]
@bleskes @clintongormley https://github.com/elastic/elasticsearch/pull/19454/commits/48c1d9d24befe945e133fbce0a40b503ed21691f contains the documentation updates. I was not sure if I should add something to the resiliency page or remove the entry in the resiliency page that talks about us adding write consistency level: https://github.com/elastic/elasticsearch/blob/master/docs/resiliency/index.asciidoc#validate-quorum-before-accepting-a-write-request-status-done-v140

File: core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java [hunk 9]
```diff
@@ -70,11 +70,6 @@ protected ReplicaResult shardOperationOnReplica(ShardFlushRequest request) {
     }
 
     @Override
-    protected boolean checkWriteConsistency() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java]: 🎈 🎉

File: core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java [hunk 10]
```diff
@@ -35,17 +36,24 @@
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.TransportService;
 
+import java.util.function.Supplier;
+
 public class TransportShardRefreshAction
         extends TransportReplicationAction<BasicReplicationRequest, BasicReplicationRequest, ReplicationResponse> {
 
     public static final String NAME = RefreshAction.NAME + "[s]";
+    private static final Supplier<BasicReplicationRequest> requestSupplier = () -> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java]: Why not set it when we translate from RefreshRequest to these shard level requests?

File: core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java [hunk 11]
```diff
@@ -63,8 +65,15 @@ public static ActiveShardCount from(final int value) {
...
+    /**
+     * Validates that the instance is valid for the given number of replicas in an index.
+     */
+    public boolean validate(final int numberOfReplicas) {
+        return value <= numberOfReplicas + 1;
+    }
+
     private static ActiveShardCount get(final int value) {
-        switch (validateValue(value)) {
+        switch (value) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java]: can we assert that the value > 1 in the default clause?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 12]
```diff
@@ -566,8 +558,8 @@ protected void doRun() {
             }
 
             // resolve all derived request fields, so we can route and apply it
-            if (request.consistencyLevel() == WriteConsistencyLevel.DEFAULT) {
-                request.consistencyLevel(defaultWriteConsistencyLevel);
+            if (request.waitForActiveShards() == ActiveShardCount.DEFAULT) {
+                request.waitForActiveShards(WAIT_FOR_ACTIVE_SHARDS_SETTING.get(indexMetaData.getSettings()));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: hmm... that does a map lookup on every op, not got. Can we have the previous variable back and add a setting listener that updates it?

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 13]
```diff
@@ -345,7 +347,11 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                                 .setRoutingNumShards(routingNumShards);
                             // Set up everything, now locally create the index to see that things are ok, and apply
                             final IndexMetaData tmpImd = tmpImdBuilder.settings(actualIndexSettings).build();
-                            if (request.waitForActiveShards().resolve(tmpImd) > tmpImd.getNumberOfReplicas() + 1) {
+                            ActiveShardCount waitForActiveShards = request.waitForActiveShards();
+                            if (waitForActiveShards == ActiveShardCount.DEFAULT) {
+                                waitForActiveShards = WAIT_FOR_ACTIVE_SHARDS_SETTING.get(tmpImd.getSettings());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: double checking - did we document this somewhere?

File: core/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 14]
```diff
@@ -127,6 +128,16 @@
     public static final Setting<Integer> MAX_SLICES_PER_SCROLL = Setting.intSetting("index.max_slices_per_scroll",
         1024, 1, Property.Dynamic, Property.IndexScope);
 
+    /**
+     * The number of active shard copies required for a write operation.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/IndexSettings.java]: please soften "required"

File: core/src/test/java/org/elasticsearch/action/support/ActiveShardCountTests.java [hunk 15]
```diff
@@ -119,15 +86,11 @@ public void testEnoughShardsActiveZero() {
         final String indexName = "test-idx";
         final int numberOfShards = randomIntBetween(1, 5);
         final int numberOfReplicas = randomIntBetween(4, 7);
-        final ActiveShardCount waitForActiveShards = ActiveShardCount.from(0);
+        final ActiveShardCount waitForActiveShards = ActiveShardCount.NONE;
         ClusterState clusterState = initializeWithNewIndex(indexName, numberOfShards, numberOfReplicas);
         assertTrue(waitForActiveShards.enoughShardsActive(clusterState, indexName));
         clusterState = startPrimaries(clusterState, indexName);
         assertTrue(waitForActiveShards.enoughShardsActive(clusterState, indexName));
-        clusterState = startLessThanWaitOnShards(clusterState, indexName, waitForActiveShards);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/ActiveShardCountTests.java]: why were these removed?

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/ActiveShardCountTests.java]: I guess it was the name of the test and these were moved.

File: core/src/test/java/org/elasticsearch/action/support/WaitActiveShardCountIT.java [hunk 16]
```diff
@@ -35,26 +34,26 @@
...
 
-        // indexing, by default, will work (ONE consistency level)
-        client().prepareIndex("test", "type1", "1").setSource(source("1", "test")).setConsistencyLevel(WriteConsistencyLevel.ONE).execute().actionGet();
+        // indexing, by default, will work (waiting for one shard copy only)
+        client().prepareIndex("test", "type1", "1").setSource(source("1", "test")).execute().actionGet();
+        ActiveShardCount activeShardCount = ActiveShardCount.from(2); // wait for two active shard copies
         try {
             client().prepareIndex("test", "type1", "1").setSource(source("1", "test"))
-                    .setConsistencyLevel(WriteConsistencyLevel.QUORUM)
+                    .setWaitForActiveShards(activeShardCount)
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/WaitActiveShardCountIT.java]: I wonder if we should a variant that gets something nummeric

File: core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java [hunk 17]
```diff
@@ -251,34 +251,16 @@ public Result perform(Request request) throws Exception {
...
-                // we have to reset default (as the transport replication action will do)
-                request.consistencyLevel(WriteConsistencyLevel.QUORUM);
-                break;
-            case ALL:
-                passesWriteConsistency = unassignedReplicas == 0;
-                break;
-            default:
-                throw new RuntimeException("unknown consistency level [" + request.consistencyLevel() + "]");
-        }
+        final int activeShardCount = randomIntBetween(0, totalShards);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java]: if this is totalShards, can we randomly use ALL?

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java [hunk 18]
```diff
@@ -254,7 +255,7 @@ private IndexShard newShard(boolean primary, DiscoveryNode node, IndexMetaData i
         public int indexDocs(final int numOfDoc) throws Exception {
             for (int doc = 0; doc < numOfDoc; doc++) {
                 final IndexRequest indexRequest = new IndexRequest(index.getName(), "type", Integer.toString(docId.incrementAndGet()))
-                    .source("{}");
+                    .source("{}").waitForActiveShards(ActiveShardCount.NONE);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: why is this?

File: core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java [hunk 19]
```diff
@@ -223,4 +229,35 @@ public void testInvalidIndexName() {
                     e.getMessage().contains("Invalid index name [..], must not be \'.\' or '..'"), equalTo(true));
         }
     }
+
+    /**
+     * This test ensures that indexing operations adhere to the {@link IndexSettings#WAIT_FOR_ACTIVE_SHARDS_SETTING}.
+     */
+    public void testChangeWaitForActiveShardsSetting() throws Exception {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java]: does this have to be an IT test? can't we do a unit test on the settings level?

File: core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java [hunk 20]
```diff
@@ -223,4 +229,35 @@ public void testInvalidIndexName() {
...
+        final int numReplicas = internalCluster().numDataNodes() + randomIntBetween(0, 3);
+        final Settings originalSettings = Settings.builder()
+                                              .put(IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), 1)
+                                              .put(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                                              .build();
+        assertAcked(client().admin().indices().prepareCreate(indexName).setSettings(originalSettings).get());
+
+        // active shard count not met, should throw exception
+        client().admin().indices().prepareUpdateSettings(indexName).setSettings(
+            Settings.builder().put(WAIT_FOR_ACTIVE_SHARDS_SETTING.getKey(), Integer.toString(internalCluster().numDataNodes() + 1)).build()
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java]: sometimes use an illegal negative value?

**Comment by bleskes** [comment]
Thx @abeyad . Left some very minor comments.

>  I was not sure if I should add something to the resiliency page or remove the entry in the resiliency page that talks about us adding write consistency level: https://github.com/elastic/elasticsearch/blob/master/docs/resiliency/index.asciidoc#validate-quorum-before-accepting-a-write-request-status-done-v140

I would just removed that paragraph. It never made it into official code - I think it was done on the improve_zen feature branch back in 1.4 but removed before the merge into master.

File: core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java (hunk 10 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java]: You're right, thanks for catching it

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java (hunk 13 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: document the `index.write.wait_for_active_shards` setting?  its included in the PR for doc updates: e.g. https://github.com/elastic/elasticsearch/pull/19581/files#diff-e973e987673cceb34ca841a82facc508R322

File: core/src/test/java/org/elasticsearch/action/support/ActiveShardCountTests.java (hunk 15 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/action/support/ActiveShardCountTests.java]: Yes, they didn't make sense for this particular test

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java (hunk 13 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: document that it has an effect on index creation. I saw it later. All good.

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java (hunk 18 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: I have no idea why that is there actually, I don't recall and it looks like remnants from some experimentation. I'm removing it.

File: core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java (hunk 17 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java]: Done

File: core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java (hunk 20 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java]: Done

File: core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java (hunk 19 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java]: I suppose it doesn't have to be, but I wanted to test not just the setting changing but that the replication operation uses the new setting... so this was the most convenient way to achieve that.  I have a similar test for index creation in `CreateIndexIT`

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java]: Two unit tests, one for replication operation the test it listens to the request and one for TransportReplicationAction that tests it resolve request based on index meta data is way better then an integration test that does way more in terms of setup and tear downs

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java (hunk 18 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: I remember now why I had to put this in place.... the `IndexingOp` which extends `ReplicationOperation` passes a cluster state `Supplier` that always gives `null`... and `ReplicationOperation#checkActiveShardCount` depends on a non-null cluster state. In this case, the better solution is to override `checkActiveShardCount` in `IndexingOp`, which is what I've done and will push shortly.

File: core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java (hunk 19 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/indexing/IndexActionIT.java]: We have these tests in `ReplicationOperationTests`, so I will add them for `TransportReplicationActionTests` and get rid of it here.

**Comment by abeyad** [comment]
@bleskes I pushed https://github.com/elastic/elasticsearch/pull/19454/commits/8f7f4bdf5bd51badfd07f89a62d6ca36ab46b01c that addresses your code review comments

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 21]
```diff
@@ -219,6 +220,16 @@ public static State fromString(String state) {
     public static final Setting<Settings> INDEX_ROUTING_INITIAL_RECOVERY_GROUP_SETTING =
         Setting.groupSetting("index.routing.allocation.initial_recovery."); // this is only setable internally not a registered setting!!
 
+    /**
+     * The number of active shard copies to check for before proceeding with a write operation.
+     */
+    public static final Setting<ActiveShardCount> WAIT_FOR_ACTIVE_SHARDS_SETTING =
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: It's a bit of mess already - so there are many conventions here - but let's try to follow of the majority here and name this `SETTING_WAIT_FOR_ACTIVE_SHARDS` ? also it will be good to open a follow up to discuss location of index settings and their naming conventions. It's all over the place now

File: core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java [hunk 22]
```diff
@@ -217,6 +217,29 @@ public static ClusterState stateWithActivePrimary(String index, boolean activePr
     }
 
     /**
+     * Creates cluster state with and index that has one shard and as many replicas as numberOfReplicas.
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java]: not sure whether the as numberOfReplicas makes sense here?

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 23]
```diff
@@ -678,6 +680,91 @@ protected ReplicaResult shardOperationOnReplica(Request request) {
...
     }
 
+    /**
+     * This test ensures that replication operations adhere to the {@link IndexMetaData#WAIT_FOR_ACTIVE_SHARDS_SETTING} setting.
+     */
+    public void testChangeWaitForActiveShardsSetting() throws Exception {
+        final String indexName = "test";
+
+        // test negative wait_for_active_shards value fails
+        expectThrows(IllegalArgumentException.class, () ->
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: this is really the job of TRA to test this? it's an index meta data thing. Whis is it here?

I think testing here should be very minimal and just check that the TRA funnel index level setting to a request object *IF* the request has it's method set to default.

**Comment by bleskes** [comment]
@abeyad thx. I left some minor comments.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java (hunk 21 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: I agree with this, I was thoroughly confused myself as to what goes in `IndexMetaData` vs `IndexSettings`, as well as the conventions. Also, some specialized classes hold their own index level settings. I will make the change to `SETTING_WAIT_FOR_ACTIVE_SHARDS` and open a discussion issue

Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: I created https://github.com/elastic/elasticsearch/issues/19723 to discuss the issue of index-level setting names and locations

File: core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java (hunk 22 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/action/support/replication/ClusterStateCreationUtils.java]: I changed the wording

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java (hunk 23 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: True, I already have negative tests for this setting, so its a bit redundant to test it with respect to TRA here. I've changed the tests accordingly.

**Comment by abeyad** [comment]
@bleskes I pushed https://github.com/elastic/elasticsearch/pull/19454/commits/81640413e56769929d7f1a85c50983bd49542bd8

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java [hunk 24]
```diff
@@ -475,4 +476,34 @@ public void testCreateShrinkIndexFails() throws Exception {
         ensureGreen();
         assertHitCount(client().prepareSearch("target").setSize(100).setQuery(new TermsQueryBuilder("foo", "bar")).get(), 20);
     }
+
+    /**
+     * This test ensures that index creation adheres to the {@link IndexMetaData#SETTING_WAIT_FOR_ACTIVE_SHARDS}.
+     */
+    public void testChangeWaitForActiveShardsSetting() throws Exception {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java]: why is this called change?

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 25]
```diff
@@ -681,84 +680,35 @@ protected ReplicaResult shardOperationOnReplica(Request request) {
     }
 
     /**
-     * This test ensures that replication operations adhere to the {@link IndexMetaData#WAIT_FOR_ACTIVE_SHARDS_SETTING} setting.
+     * This test ensures that replication operations adhere to the {@link IndexMetaData#SETTING_WAIT_FOR_ACTIVE_SHARDS} setting
+     * when the request is using the default value for waitForActiveShards.
      */
     public void testChangeWaitForActiveShardsSetting() throws Exception {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: why is this called change? we don't change anythig do we?

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 26]
```diff
@@ -681,84 +680,35 @@ protected ReplicaResult shardOperationOnReplica(Request request) {
...
+        assertEquals(ActiveShardCount.from(idxSettingWaitForActiveShards), request.waitForActiveShards());
+
+        // test wait_for_active_shards when default not set on the request (request value should be honored over index setting)
+        int requestWaitForActiveShards = randomIntBetween(0, numReplicas + 1);
+        request = new Request(shardId).waitForActiveShards(ActiveShardCount.from(requestWaitForActiveShards));
+        action.resolveRequest(state.metaData(), state.metaData().index(indexName), request);
+        assertEquals(ActiveShardCount.from(requestWaitForActiveShards), request.waitForActiveShards());
     }
 
     private ClusterState changeWaitForActiveShardsSetting(String indexName, ClusterState state, int waitForActiveShards) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: can we inline this - it is ionly used once

**Comment by bleskes** [comment]
thx @abeyad . Almost there. Left some very minor comments.

File: core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java (hunk 24 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java]: Not a great name, I'll change it to something more descriptive like `testDefaultWaitForActiveShardsUsesIndexSetting` and for the other one.

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java (hunk 26 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: sure

**Comment by abeyad** [comment]
@bleskes i pushed https://github.com/elastic/elasticsearch/pull/19454/commits/4923da93c83b0b0708747105f427aab03047168b

**Comment by bleskes** [comment]
LGTM. Thx @abeyad

**Comment by abeyad** [comment]
Thanks for your review and feedback @bleskes !
